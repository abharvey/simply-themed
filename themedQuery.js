/**
 * usage:
 * 
 * 
 *  const sizeTheme = {
 *      small: {
 *         break: { max: 1024 },
 *         spacing: {
 *              extraSmall: 4,
 *              small: 8,
 *              medium: 16,
 *              large: 20,
 *              extraLarge: 24
 *         },
 *          font: {
 *              small: 1.25rem,
 *              normal: 1.5rem,
 *              strong: 2rem,
 *          }
 *      },
 *      medium: {
 *          break: { min: 1023, max: 1366 },
 *          spacing: {
 *               extraSmall: 4,
 *               small: 8,
 *               medium: 12,
 *               large: 16,
 *               extraLarge: 20
 *          },
 *          font: {
 *              small: 1rem,
 *              normal: 1.25rem,
 *              strong: 1.5rem,
 *          }
 *      },
 *      large: {
 *          break: { min: 1367 },
 *          spacing: {
 *              extraSmall: 4,
 *              small: 8,
 *              medium: 10,
 *              large: 12,
 *              extraLarge: 16
 *          },
 *          font: {
 *              small: .8125rem,
 *              normal: 1rem,
 *              strong: 1.25rem,
 *          }
 *      }
 *  };
 * 
 * const sizing = themedBreaks(sizeTheme);
 * 
 * const StyledDiv = styled.div`
 *      ${theme`
 *          padding: ${spacing.small}px ${spacing.medium}px;
 *          margin: ${spacing.large}px ${spacing.extraLarge}px;
 *          font-size: ${font.normal};
 *      `}
 * `;
 * 
 */

import mediaQueryMixin from './breakpoints';

// TODO: work on the naming

const weaveCSS = (cssStrings, cssArgs) => {
    const result = [cssStrings[0]];

    for (let i = 0, len = cssArgs.length; i < len; i += 1) {
        result.push(cssArgs[i], cssStrings[i + 1]);
    }
    // single css rule string ie 'padding: 4px 12px;'
    return result.reduce((str, arg) => `${str}${arg}`, '');
};

const buildBreakPointThemeMap = (theme, subThemeKey, cssArgs) => {
    const themeSizes = Object.keys(theme);

    const subTheme = themeSizes.reduce((subThemeMap, sizeKey) => {
        subThemeMap[sizeKey] = theme[sizeKey][subThemeKey];
        return subThemeMap; //{ small: {'spacing': { small, medium, large } }}
    }, {})

    return themeSizes.reduce((breakPointSizeMap, size) => {
        breakPointSizeMap[size] = cssArgs.map((cssArg) => subTheme[size][cssArg]);
        return breakPointSizeMap;
    }, {});
};

// SubTheme reduction should be done in another function at the same time
//  as the theme size reduction

export default (cssFunc) => (theme, subThemeKey) => (cssStrings, ...cssArgs) => {
    // create the breakpoint media queries from the theme (pull out the break object)    
    // format should be theme = { key: { break: { min, max } } }
    const themeSizes = Object.keys(theme);

    const breakpoints = mediaQueryMixin(cssFunc)(themeSizes.reduce((bpm, sizeKey) => {
        bpm[sizeKey] = theme[sizeKey].break;
        return bpm; //{ small: {max}, medium: {min, max}, large: {min}}}
    }, {})); //{ small: func, medium: func, large: func}

    // converts the arguments to the sub theme style, ie: ${'small'} might map to 8 on a medium
    const breakPointTheme = buildBreakPointThemeMap(theme, subThemeKey, cssArgs);

    const css = themeSizes.map(size => {
        return cssFunc`${breakpoints[size]`
            ${weaveCSS(cssStrings, breakPointTheme[size])}
        `}`;
    });

    return cssFunc`${css}`;
};
