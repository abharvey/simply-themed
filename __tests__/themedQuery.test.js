import * as emotion from 'emotion';
import { createSerializer, getStyles } from 'jest-emotion';
import 'jest-styled-components';
import eStyled, { css as eCSS } from 'react-emotion';
import renderer from 'react-test-renderer';
import scStyled, { css as scCSS } from 'styled-components';

// expect.extend(createMatchers(emotion));

import { cleanCSS } from './testUtils';
import themedQuery from '../src/themedQuery';

const theme = {
    small: {
        break: { max: 1024 },
        spacing: {
            extraSmall: 4,
            small: 8,
            medium: 16,
            large: 20,
            extraLarge: 24,
        },
        font: {
            small: '1.25rem',
            normal: '1.5rem',
            strong: '2rem',
        },
    },
    medium: {
        break: { min: 1023, max: 1366 },
        spacing: {
            extraSmall: 4,
            small: 8,
            medium: 12,
            large: 16,
            extraLarge: 20,
        },
        font: {
            small: '1rem',
            normal: '1.25rem',
            strong: '1.5rem',
        },
    },
    large: {
        break: { min: 1367 },
        spacing: {
            extraSmall: 4,
            small: 8,
            medium: 10,
            large: 12,
            extraLarge: 16,
        },
        font: {
            small: '.8125rem',
            normal: '1rem',
            strong: '1.25rem',
        },
    },
};

describe('Themed Media Query Function with styled-components', () => {
    let spacingQueries;
    beforeEach(() => {
        spacingQueries = themedQuery(scCSS)(theme, 'spacing');
    });

    it('should create a query function', () => {
        expect(typeof spacingQueries).toBe('function');
    });

    it('function should create multiple media queries', () => {
        const mediaQueries = spacingQueries`
                padding: ${'extraSmall'}px ${'large'}px;
            `;

        const mq = cleanCSS(mediaQueries.join('\n'));

        const expected = `
            @media (max-width:1024px) {
                padding: 4px 20px;
            }
            @media all and (max-width: 1366px) and (min-width: 1023px) {
                padding: 4px 16px;
            }
            @media (min-width: 1367px) {
                padding: 4px 12px;
            }
        `;

        expect(mq).toEqual(cleanCSS(expected));
    });

    it('should works for different subThemes', () => {
        const fontQueries = themedQuery(scCSS)(theme, 'font');

        const mediaQueries = fontQueries`
                font-size: ${'normal'};
            `;

        const mq = cleanCSS(mediaQueries.join('\n'));

        const expected = `
            @media (max-width:1024px) {
                font-size: 1.5rem;
            }
            @media all and (max-width: 1366px) and (min-width: 1023px) {
                font-size: 1.25rem;
            }
            @media (min-width: 1367px) {
                font-size: 1rem;
            }
        `;

        expect(mq).toEqual(cleanCSS(expected));
    });

    it('should work to create a react component with styled-components', () => {
        const React = require('react');
        const Test = scStyled.div`
            ${spacingQueries`
                margin: ${'large'}px ${'medium'}px;
            `}
        `;

        const tree = renderer.create(<Test />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Themed Media Query Function works with emotion', () => {
    expect.addSnapshotSerializer(createSerializer(emotion));

    let spacingQueries;
    beforeEach(() => {
        spacingQueries = themedQuery(eCSS)(theme, 'spacing');
    });

    it('should create a query function', () => {
        expect(typeof spacingQueries).toBe('function');
    });

    it('should work to create a css class with emotion', () => {
        const mediaQueries = spacingQueries`
                padding: ${'extraSmall'}px ${'large'}px;
                margin: ${'small'}px ${'extraLarge'}px;
                width: calc(${'large'}px * 2);
            `;

        const mq = cleanCSS(mediaQueries);

        expect(getStyles(emotion)).toMatchSnapshot();
    });

    it('should work to create a react component with emotion components', () => {
        const React = require('react');
        const Test = eStyled('div')`
            ${spacingQueries`
                margin: ${'large'}px ${'medium'}px;
            `}
        `;

        const tree = renderer.create(<Test />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
