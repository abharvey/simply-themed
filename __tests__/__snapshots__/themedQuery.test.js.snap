// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themed Media Query Function with styled-components should work to create a react component with styled-components 1`] = `
@media (max-width:1024px) {
  .c0 {
    margin: 20px 16px;
  }
}

@media all and (max-width:1366px) and (min-width:1023px) {
  .c0 {
    margin: 16px 12px;
  }
}

@media (min-width:1367px) {
  .c0 {
    margin: 12px 10px;
  }
}

<div
  className="c0"
/>
`;

exports[`Themed Media Query Function works with emotion should work to create a css class with emotion 1`] = `".css-hcw3g8{padding:4px 20px;margin:8px 24px;width:calc(20px * 2);}@media (max-width:1024px){.css-1h251yx{padding:4px 20px;margin:8px 24px;width:calc(20px * 2);}}.css-1h1kxvq{padding:4px 16px;margin:8px 20px;width:calc(16px * 2);}@media all and (max-width:1366px) and (min-width:1023px){.css-1glhxsy{padding:4px 16px;margin:8px 20px;width:calc(16px * 2);}}.css-1lkvvyw{padding:4px 12px;margin:8px 16px;width:calc(12px * 2);}@media (min-width:1367px){.css-1kmhy8h{padding:4px 12px;margin:8px 16px;width:calc(12px * 2);}}@media (max-width:1024px){.css-rdocqf{padding:4px 20px;margin:8px 24px;width:calc(20px * 2);}}@media all and (max-width:1366px) and (min-width:1023px){.css-rdocqf{padding:4px 16px;margin:8px 20px;width:calc(16px * 2);}}@media (min-width:1367px){.css-rdocqf{padding:4px 12px;margin:8px 16px;width:calc(12px * 2);}}"`;

exports[`Themed Media Query Function works with emotion should work to create a react component with emotion components 1`] = `
<div
  className="emotion-0"
/>
`;
